### PR TITLE
[Operators] vector_norm and var_mean

### PR DESCRIPTION
**Pull request by StrongSpoon**
1. vector_norm: support all legal inputs
    - ord: 2, inf, -inf, 0, other float/int
    - dim: None, list of int
    - keepdim: True, False
    - dtype: float16, float32, bfloat16
2. var_mean: support all legal inputs
    - dim: None, list of int
    - correction: None, float/int
    - keepdim: True, False
3. group_norm by Adam
    - forward finished
    - little error with bf16 weight grad
4. mean_dim: support all legal inputs
    - dim: None, list of int
    - keepdim: True, False

### CODE REVIEW

File: src/flag_gems/vector_norm.py [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+        shape = list(x.shape)
+        N = 1
+        order = list(range(x.ndim))
+        for i in dim:
+            i = i % x.ndim
+            order.remove(i)
+            order.append(i)
+            N *= shape[i]
+            shape[i] = 1
+        x = x.permute(order)
```
Comment (iclementine) [inline src/flag_gems/vector_norm.py]: I would suggest construct the permutation in a more efficient way, since list.remove involves comparing the value to remove with values in the list. 

In this case, you can just pop the value at a specific index and append it at the back.

```python
order.append(order.pop(i))
```

or just construct the permutation a-fresh without init-and-modify.

```python
order = (i for i in range(x.dim) if i not in dim) + tuple(dim)
```

BTW, should dim be sorted beforehand to avoid harder permutation?

File: src/flag_gems/vector_norm.py [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+        mask = row_mask and col_mask
+
+        a = tl.load(X + cols, mask, other=0.0).to(tl.float32)
+        _sum += tl.math.pow(tl.abs(a), ord)
+    sum = tl.sum(_sum, axis=1)
+    out = tl.math.pow(sum, 1 / ord)[:, None]
+    tl.store(Out, out, row_mask)
+
+
+def vector_norm(x, ord=2, dim=None, keepdim=False, dtype=None):
```
Comment (iclementine) [inline src/flag_gems/vector_norm.py:128]: `dim` should also accept int, in addition to tuple or list of int, according to pytorch convention.

```python
import torch
import flag_gems
x = torch.randn(1, 1, 16, 1, device="cuda")
flag_gems.vector_norm(x, ord=2, dim=2, keepdim=False)
```

File: src/flag_gems/vector_norm.py [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+    elif ord == float("inf"):
+        max_norm_kernel[grid](x, out, M, N)
+    elif ord == -float("inf"):
+        min_norm_kernel[grid](x, out, M, N)
+    elif ord == 0:
+        l0_norm_kernel[grid](x, out, M, N)
+    else:
+        v_norm_kernel[grid](x, out, M, N, ord)
+    if not keepdim:
+        out = out.squeeze()
```
Comment (iclementine) [inline src/flag_gems/vector_norm.py]: When not `keepdim` only the axis specified in dim should be squeeze out. Or else, other axis with size equals 1 would also be squeezed out.

Use the following snippet to test

```python
import torch
import flag_gems
x = torch.randn(1, 1, 16, 1, device="cuda")
print(torch.linalg.vector_norm(x, ord=2, dim=(2, ), keepdim=False).shape)
print(flag_gems.vector_norm(x, ord=2, dim=(2, ), keepdim=False).shape)
```

File: src/flag_gems/vector_norm.py [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+@libentry()
+@triton.autotune(configs=cfggen(), key=["M", "N"])
+@triton.jit
+def l0_norm_kernel(X, Out, M, N, BLOCK_M: tl.constexpr, BLOCK_N: tl.constexpr):
+    pid = tl.program_id(0) * BLOCK_M + tl.arange(0, BLOCK_M)[:, None]
+    X = X + pid * N
+    Out = Out + pid
+    row_mask = pid < M
+
+    _sum = tl.zeros([BLOCK_M, BLOCK_N], dtype=tl.float32)
```
Comment (iclementine) [inline src/flag_gems/vector_norm.py:93]: I have a concern about data type here. What if the input data is double, then the accumulator type should at least be double. What sbout adding a tl.constexpr parameter `acc_type`?

File: src/flag_gems/vector_norm.py (hunk 2 above)
Comment (StrongSpoon) [reply src/flag_gems/vector_norm.py:128]: integer dim is precessed as a list of length 1

File: src/flag_gems/vector_norm.py (hunk 4 above)
Comment (StrongSpoon) [reply src/flag_gems/vector_norm.py:93]: this is a common problem for all operators. there are too many expressions using fixed datatype, which is unfriendly for datatype expansion.

File: src/flag_gems/var_mean.py [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+        dim = 0
+        M = 1
+        N = x.numel()
+        shape = [1] * x.ndim
+        x = x.flatten()
+    else:
+        shape = list(x.shape)
+        N = 1
+        order = list(range(x.ndim))
+        for i in dim:
```
Comment (iclementine) [inline src/flag_gems/var_mean.py]: same case with vector_norm

File: src/flag_gems/vector_norm.py (hunk 2 above)
Comment (iclementine) [reply src/flag_gems/vector_norm.py:128]: Yes, but 
图片

File: src/flag_gems/vector_norm.py (hunk 3 above)
Comment (iclementine) [reply src/flag_gems/vector_norm.py]: What about just unsqueeze the reduce axes?

File: src/flag_gems/vector_norm.py (hunk 2 above)
Comment (iclementine) [reply src/flag_gems/vector_norm.py:128]: I see, flag_gems.vector_norm is not called directly. It is called by overriding aten::linalg_vector_norm. 
Resolved.

File: src/flag_gems/vector_norm.py (hunk 1 above)
Comment (StrongSpoon) [reply src/flag_gems/vector_norm.py]: done. I think it doesn't matter whether dim is sorted or not, cause input tensor doesn't have to be contiguous.

File: src/flag_gems/vector_norm.py (hunk 3 above)
Comment (StrongSpoon) [reply src/flag_gems/vector_norm.py]: done

File: src/flag_gems/var_mean.py (hunk 5 above)
Comment (StrongSpoon) [reply src/flag_gems/var_mean.py]: done

File: src/flag_gems/vector_norm.py (hunk 1 above)
Comment (iclementine) [reply src/flag_gems/vector_norm.py]: Yes, the input tensor may not be contiguous.

File: src/flag_gems/groupnorm.py [hunk 6]
```diff
@@ -0,0 +1,317 @@
...
+        Y_val += bias
+
+    tl.store(Y_block_ptr, Y_val.to(X_val.dtype), boundary_check=(0, 1))
+    tl.store(rstd_block_ptr, rstd.to(X_val.dtype))
+    tl.store(mean_block_ptr, mean.to(X_val.dtype))
+
+
+@libentry()
+@triton.jit
+def group_norm_backward_kernel(
```
Comment (iclementine) [inline src/flag_gems/groupnorm.py:118]: The backward kernel is launched with grid `(N * num_groups, 1, 1)`. Then the gradient of W & B should not be computed like this.

A block of the kernel loads a block of size `(BLOCK_GROUP_SIZE,)` from W or B `(size: num_groups, group_size)`， while  X, Y & Z(normalized X) have shape (N, num_groups, group_size, H, W).

The gradien of W should be the gradient of Y reduced sum by axes( N, H, W). In a task partitioning with grid `(N* num_groups, 1, 1`, it would require a reduce across blocks to accompolish the reduction.

I suggest the use a different task partitioning scheme for computing the gradient of W & B.
